File: java/454.4-sum-ii.java
/*
 * @lc app=leetcode id=454 lang=java
 *
 * [454] 4Sum II
 */
class Solution {
    public int fourSumCount(int[] A, int[] B, int[] C, int[] D) {
        int res = 0;
        Map<Integer, Integer> map = new HashMap<>();
        for (int d = 0; d < D.length; ++d) {
            for (int c = 0; c < C.length; ++c) {
                map.put(-C[c]-D[d], map.getOrDefault(-C[c]-D[d], 0) + 1);
            }
        }
        for (int b = 0; b < B.length; ++b) {
            for (int a = 0; a < A.length; ++a) {
                if (map.containsKey(A[a]+B[b]))
                    res += map.get(A[a]+B[b]);
            }
        }

        return res;
    }
}

File: go/454.4-sum-ii.go
/*
 * @lc app=leetcode id=454 lang=golang
 *
 * [454] 4Sum II
 */

// @lc code=start
func fourSumCount(A []int, B []int, C []int, D []int) int {
	// O(n^2)
	cnt := make(map[int]int)
	for _, va := range A {
		for _, vb := range B {
			if v, b := cnt[va+vb]; b {
				cnt[va+vb] = v + 1
			} else {
				cnt[va+vb] = 1
			}
		}
	}
	res := 0
	for _, vc := range C {
		for _, vd := range D {
			if v, b := cnt[-vc-vd]; b {
				res += v
			}
		}
	}
	return res
}

// @lc code=end

